packages/plugin-rss/src/plugin-rss.ts

```diff
@@ -0,0 +1,153 @@
+import { createFeed, generateFeedItem } from './feed';
+
+import {
+  PageWithFeeds,
+  ResolvedOutput,
+  concatArray,
+  writeFile,
+} from './internals';
+import { getDefaultFeedOption, getOutputInfo, testPage } from './options';
+import type { FeedChannel, FeedItem, PluginRssOptions } from './type';
+
+type FeedItemWithChannel = FeedItem & { channel: string };
+type TransformedFeedChannel = FeedChannel & { output: ResolvedOutput };
+
+class FeedsSet {
+  feeds: TransformedFeedChannel[] = [];
+  feedsMapById: Record<string, TransformedFeedChannel> = Object.create(null);
+  set({ feed, output, siteUrl }: PluginRssOptions, config: UserConfig) {
+    this.feeds = (
+      Array.isArray(feed) ? feed : [{ ...getDefaultFeedOption(), ...feed }]
+    ).map(options => ({
+      title: config.title || '',
+      description: config.description || '',
+      favicon: config.icon && resolveUrl(siteUrl, config.icon),
+      copyright: config.themeConfig?.footer?.message || '',
+      link: siteUrl,
+      docs: '',
+      ...options,
+      output: getOutputInfo(options, { siteUrl, output }),
+    }));
+
+    this.feedsMapById = this.feeds.reduce(
+      (m, f) => ({ ...m, [f.id]: f }),
+      Object.create(null),
+    );
+  }
+  get(): TransformedFeedChannel[];
+  get(id: string): TransformedFeedChannel | null;
+  get(id?: string): TransformedFeedChannel[] | TransformedFeedChannel | null {
+    if (id) {
+      return this.feedsMapById[id] || null;
+    }
+    return this.feeds.slice(0);
+  }
+}
+
+function getRssItems(
+  feeds: TransformedFeedChannel[],
+  page: PageIndexInfo,
+  config: UserConfig,
+  siteUrl: string,
+): Promise<FeedItemWithChannel[]> {
+  return Promise.all(
+    feeds
+      .filter(options => testPage(options.test, page, config.base))
+      .map(async options => {
+        const after = options.item || ((feed: FeedItem) => feed);
+        const item = await after(
+          generateFeedItem(page, siteUrl),
+          page,
+          siteUrl,
+        );
+        return { ...item, channel: options.id };
+      }),
+  );
+}
+
+export function pluginRss(pluginRssOptions: PluginRssOptions): RspressPlugin {
+  const feedsSet = new FeedsSet();
+
+  /**
+   * workaround for retrieving data of pages in `afterBuild`
+   * TODO: get pageData list directly in `afterBuild`
+   **/
+  let _rssWorkaround: null | Record<string, FeedItemWithChannel[]> = null;
+  let _config: null | UserConfig;
+
+  return {
+    name: PluginName,
+    globalUIComponents: Object.values(PluginComponents),
+    beforeBuild(config, isProd) {
+      if (!isProd) {
+        _rssWorkaround = null;
+        return;
+      }
+      _rssWorkaround = {};
+      _config = config;
+      feedsSet.set(pluginRssOptions, config);
+    },
+    async extendPageData(_pageData) {
+      if (!_rssWorkaround) return;
+
+      const pageData = _pageData as PageWithFeeds;
+      // rspress run `extendPageData` twice for each page - we need one only
```

Comment:
That's because in production we have two build process——client build and ssg build, which leads to the twice execution as you see, but it's neccessary to keep the same page data for client render and ssg render process.

Reply:
comment fixed - actually the extended page data with rss is for both.